Occupancy bed tooltip fully visible in mobile view

Rishith25 opened the pull request:
## Proposed Changes

- Fixes #9230 
- Occupancy bed tooltip fully visible in mobile view

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a tooltip for displaying occupancy information on the Facility Card.
	- Enhanced interactivity with the tooltip's visibility controlled by button clicks.
	- Streamlined tooltip integration in the Sidebar component for improved usability.
	- Updated tooltip implementation to utilize a new `TooltipComponent` for better functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Rishith25 committed:
occupancy bed tooltip fully visible in mobile view

rithviknishad reviewed (changes requested):
Let's switch to newer approach of showing tooltip using shadcn's tooltip component: https://github.com/ohcnetwork/care_fe/blob/develop/src/components/ui/tooltip.tsx

Rishith25 commented:
> Let's switch to newer approach of showing tooltip using shadcn's tooltip component: https://github.com/ohcnetwork/care_fe/blob/develop/src/components/ui/tooltip.tsx

This shadcn tooltip component working well in desktop, but it is not working on mobile view how should I proceed?

rithviknishad commented:
Could you push the changes so that we can take a look at it?

Rishith25 committed:
Used Tooltip in Occupancy

Rishith25 commented:
> Could you push the changes so that we can take a look at it?

Could you please look into the changes

Rishith25 committed:
Merge branch 'develop' into occupany-bed-tooltip

rithviknishad commented:
I don't see any issues in this. What's the issue?

Rishith25 commented:
> I don't see any issues in this. What's the issue?

In mobile view when we are clicking on the occupancy, the tooltip is appearing and disappearing quickly.
it is happening only in mobile view

rithviknishad commented:
https://github.com/shadcn-ui/ui/issues/86#issuecomment-1472176666

Rishith25 committed:
Merge branch 'develop' into occupany-bed-tooltip

Rishith25 committed:
Using tooltip for facility card occupancy

Rishith25 commented:
@rithviknishad  Now it is working fine in mobile view could you please look into it?

rithviknishad commented on src/components/Facility/FacilityCard.tsx:
```diff
@@ -98,19 +106,38 @@ export const FacilityCard = (props: {
+                                }`}
+                                onClick={() => setOpen(!open)}
+                                onMouseEnter={() => setOpen(true)}
+                                onMouseLeave={() => setOpen(false)}
```
The badge is not a button

rithviknishad replied:
this change should be made to the tooltip component instead.

rithviknishad commented on src/components/Facility/FacilityCard.tsx:
```diff
@@ -98,19 +106,38 @@ export const FacilityCard = (props: {
+                            <TooltipContent
+                              side="top"
+                              sideOffset={8}
+                              className="z-50 max-w-[90%] break-words text-xs text-center sm:max-w-xs md:max-w-md lg:max-w-lg xl:max-w-xl bg-gray-900 text-gray-50 rounded-md p-2 shadow-lg animate-in fade-in-0 zoom-in-95 dark:bg-gray-50 dark:text-gray-900"
```
it's responsive without any classes right?
```suggestion
```

Rishith25 committed:
Added changes of onMouse in tooltip

rithviknishad commented on src/components/ui/tooltip.tsx:
```diff
@@ -12,17 +13,22 @@ const TooltipTrigger = TooltipPrimitive.Trigger;
-  />
-));
+>(({ className, sideOffset = 4, ...props }, ref) => {
+  const [_open, setOpen] = useState(false);
```
if we are not using the `_open` state, why do we have this state in the first place?

rithviknishad replied:
@coderabbitai does this make sense?

Rishith25 committed:
Tooltip updated

rithviknishad commented:
@Rishith25 do a self review of the code

Rishith25 committed:
Merge branch 'develop' into occupany-bed-tooltip

Rishith25 commented:
> @Rishith25 do a self review of the changes

@rithviknishad  In my recent commit I have removed the unused `_open`. 
As Tooltip clicks are not supported in mobile view I have done the changes according to this https://github.com/shadcn-ui/ui/issues/86#issuecomment-2241817826 which you have suggested and used the `onMouseEnter` and `onMouseLeave` in `tooltip.tsx` file and handled click Event in `FacilityCard` for mobile view

rithviknishad commented on src/components/ui/tooltip.tsx:
```diff
@@ -12,17 +12,21 @@ const TooltipTrigger = TooltipPrimitive.Trigger;
+        "z-50 overflow-hidden rounded-md bg-gray-900 px-3 py-1.5 text-xs text-gray-50 animate-in fade-in-0 zoom-in-95 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 dark:bg-gray-50 dark:text-gray ...
+        className,
+      )}
+      onMouseEnter={() => true}
+      onMouseLeave={() => false}
```
are you sure this does anything?

Rishith25 replied:
@rithviknishad  Yeah I removed it and its still working fine so committing changes

Rishith25 committed:
Updated tooltip for occupancy badge

Rishith25 committed:
Merge branch 'develop' into occupany-bed-tooltip

Rishith25 committed:
Tooltip for occupancy in facility card

Rishith25 commented:
Hey @rithviknishad, the tooltip is now working fine on both mobile and desktop. I have added a click event in `FacilityCard` for mobile view and removed `onMouseEnter` and `onMouseLeave` as they are not needed. Even without them, the tooltip works perfectly on mobile using the `onClick` event. Could you please review it?

Jacobjeevan commented on src/components/Facility/FacilityCard.tsx:
```diff
@@ -19,6 +19,13 @@ import routes from "@/Utils/request/api";
+  TooltipContent,
+  TooltipProvider,
+  TooltipTrigger,
+} from "../ui/tooltip";
```
Replace w/ absolute imports

Rishith25 committed:
Merge branch 'develop' into occupany-bed-tooltip

Rishith25 committed:
Toottip absolute imports in Facility Card

nihal467 commented:
LGTM @Jacobjeevan can you review it

rithviknishad commented on src/components/Facility/FacilityCard.tsx:
```diff
@@ -29,6 +36,7 @@ export const FacilityCard = (props: {
   const [notifyModalFor, setNotifyModalFor] = useState<string>();
   const [notifyMessage, setNotifyMessage] = useState("");
   const [notifyError, setNotifyError] = useState("");
+  const [open, setOpen] = useState(false);
```
The naming for this state is not appropriate here. state `open` in the context of FacilityCard has an ambiguous meaning. 

Also, like already said in one of the previous reviews, keep the state inside the tooltip component itself. Since this behaviour is expected in all usages of the tooltip and not just the FacilityCard.

Rishith25 committed:
Merge branch 'develop' into occupany-bed-tooltip

Rishith25 commented:
Hi @rithviknishad  
As per your suggestion I tried to include onClick in the TooltipTrigger  of tooltip.tsx and it showed some unexpected behavior and further optimized it by Creating a tooltip component `TooltipComponent` in tooltip.tsx file and this tooltip component can be used directly in the facility card or any other components. So if this is okay I will also update the other components which are using this such as Slidebar.tsx
  
`FacilityCard.tsx`
```.js
<TooltipProvider>
  <TooltipComponent content={t("live_patients_total_beds")}>
    <div
      className={`relative flex items-center gap-1 text-sm ${
        (facility.patient_count || 0) / (facility.bed_count || 0) > 0.85
          ? "justify-center rounded-md border border-red-600 bg-red-500 p-1 font-bold text-white"
          : "text-secondary-700"
      }`}
    >
      <CareIcon icon="l-bed" />
      <dt>
        {t("occupancy")}: {facility.patient_count} / {facility.bed_count}
      </dt>
    </div>
  </TooltipComponent>
</TooltipProvider>;

```

`tooltip.tsx`
```.js
import * as TooltipPrimitive from "@radix-ui/react-tooltip";
import * as React from "react";

import { cn } from "@/lib/utils";

const TooltipProvider = TooltipPrimitive.Provider;

const Tooltip = TooltipPrimitive.Root;

const TooltipTrigger = TooltipPrimitive.Trigger;

const TooltipContent = TooltipPrimitive.Content;

const TooltipComponent = React.forwardRef<
  React.ElementRef<typeof TooltipContent>,
  React.ComponentPropsWithoutRef<typeof TooltipPrimitive.Content>
>(({ children, content, sideOffset = 4, className }, ref) => {
  const [open, setOpen] = React.useState(false);

  return (
    <TooltipProvider>
      <Tooltip open={open} onOpenChange={setOpen} delayDuration={0}>
        <TooltipTrigger asChild onClick={() => setOpen(!open)}>
          {children}
        </TooltipTrigger>
        <TooltipContent
          ref={ref}
          sideOffset={sideOffset}
          className={cn(
            "z-50 overflow-hidden rounded-md bg-gray-900 px-3 py-1.5 text-xs text-gray-50 animate-in fade-in-0 zoom-in-95 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 dark:bg-gray-50 dark:text-gray-900",
            className,
          )}
        >
          {content}
        </TooltipContent>
      </Tooltip>
    </TooltipProvider>
  );
});

export { TooltipComponent, TooltipTrigger, TooltipContent, TooltipProvider };
```

Rishith25 committed:
Updated the Tooltip component file

Rishith25 committed:
Merge branch 'develop' into occupany-bed-tooltip

nihal467 commented:
@Rishith25 clear the merge conflict 

PR looks good for me, @rithviknishad @khavinshankar review it

Rishith25 commented:
> @Rishith25 clear the merge conflict
> 
> PR looks good for me, @rithviknishad @khavinshankar review it

I have changed the tooltip component so could you please confirm is it fine or not.

rithviknishad commented on src/components/Facility/FacilityCard.tsx:
```diff
@@ -107,36 +101,25 @@ export const FacilityCard = (props: {
+                              }`}
                             >
-                              <div
-                                data-test-id="occupancy-badge"
```
do resolve merge conflicts properly by trying to understand what caused the merge conflict in the first place. this was recently added in #9145

cc: @nihal467

Rishith25 committed:
Tooltip Component created for Facility Card

Rishith25 committed:
Tooltip Component created for occupancy capacity in facility

Rishith25 commented:
> > @Rishith25 clear the merge conflict
> > PR looks good for me, @rithviknishad @khavinshankar review it
> 
> I have changed the tooltip component so could you please confirm is it fine or not.

@nihal467 @rithviknishad Can you please review it and all the test cases are passing.

rithviknishad commented on src/components/ui/tooltip.tsx:
```diff
@@ -9,20 +9,39 @@ const Tooltip = TooltipPrimitive.Root;
-TooltipContent.displayName = TooltipPrimitive.Content.displayName;
+>(({ children, content, sideOffset = 4, className }, ref) => {
+  const [open, setOpen] = React.useState(false);
+  //Reusable tooltip with dynamic content.
```
```suggestion
```

rithviknishad committed:
Update src/components/ui/tooltip.tsx
